### PR TITLE
fix(storage): recreate stream when meet IO error

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/storage/src/hummock/compactor/iterator.rs [hunk 1]
```diff
@@ -30,19 +31,22 @@ use crate::hummock::compactor::task_progress::TaskProgress;
 use crate::hummock::iterator::{Forward, HummockIterator};
 use crate::hummock::sstable_store::{BlockStream, SstableStoreRef};
 use crate::hummock::value::HummockValue;
-use crate::hummock::{Block, BlockHolder, BlockIterator, HummockResult};
+use crate::hummock::{BlockHolder, BlockIterator, HummockResult, TableHolder};
 use crate::monitor::StoreLocalStatistic;
 
+const MAX_RETRY_TIME: u64 = 600; // 10min
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/iterator.rs:37]: I think we define more and more constants , it is best to manage them uniformly (another PR)

File: src/storage/src/hummock/compactor/iterator.rs [hunk 2]
```diff
@@ -132,31 +151,47 @@ impl SstableStreamIterator {
...
+            loop {
+                let now = Instant::now();
+                let ret = match &mut self.block_stream {
+                    Some(block_stream) => block_stream.next().await,
+                    None => {
+                        self.create_stream().await?;
+                        continue;
+                    }
+                };
+                let add = (now.elapsed().as_secs_f64() * 1000.0).ceil();
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/iterator.rs:164]: What does add 1000 sec mean?

Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/iterator.rs:164]: to milliseconds

File: src/storage/src/hummock/compactor/iterator.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/iterator.rs:37]: ok

File: src/storage/src/hummock/compactor/iterator.rs [hunk 3]
```diff
@@ -30,19 +31,20 @@ use crate::hummock::compactor::task_progress::TaskProgress;
...
 use crate::hummock::value::HummockValue;
-use crate::hummock::{Block, BlockHolder, BlockIterator, HummockResult};
+use crate::hummock::{BlockHolder, BlockIterator, HummockResult, TableHolder};
 use crate::monitor::StoreLocalStatistic;
 
 /// Iterates over the KV-pairs of an SST while downloading it.
-struct SstableStreamIterator {
+pub struct SstableStreamIterator {
+    sstable_store: SstableStoreRef,
+    table_holder: TableHolder,
```
Comment (zwang28) [inline src/storage/src/hummock/compactor/iterator.rs:40]: The table_holder seems a refactor/optimization that is orthogonal to recreating stream. I think it helps reduce IO but risk OOM if SST metadata is large.

I'd +1 for this refactor. @Li0k

File: src/storage/src/hummock/compactor/compactor_runner.rs [hunk 4]
```diff
@@ -142,6 +142,11 @@ impl CompactorRunner {
         task_progress: Arc<TaskProgress>,
     ) -> HummockResult<impl HummockIterator<Direction = Forward>> {
         let mut table_iters = Vec::new();
+        let compact_io_retry_time = self
+            .compactor
+            .context
+            .storage_opts
+            .object_store_streaming_read_timeout_ms;
```
Comment (zwang28) [inline src/storage/src/hummock/compactor/compactor_runner.rs:149]: object_store_streaming_read_timeout_ms is already used by a more fine-grained operation. I suggest another new option.

**Review by zwang28 — APPROVED** [review]
Rest LGTM.

File: src/storage/src/hummock/compactor/compactor_runner.rs (hunk 4 above)
Comment (Li0k) [reply src/storage/src/hummock/compactor/compactor_runner.rs:149]: +1

File: src/storage/src/hummock/compactor/iterator.rs (hunk 3 above)
Comment (Li0k) [reply src/storage/src/hummock/compactor/iterator.rs:40]: Estimated size might cover this, but when the sst meta takes up most of the space of the entire object (> estimated size), then oom can happen.

I don't have any other ideas about this other than reducing max_key_count to reduce the number of tombstones per meta. cc @zwang28 @Little-Wallace

**Review by Li0k — APPROVED** [review]
Rest LGTM

File: src/storage/src/hummock/sstable_store.rs [hunk 5]
```diff
@@ -492,37 +492,41 @@ impl SstableStore {
         self.meta_cache.get_memory_usage() as u64
     }
 
-    pub async fn get_stream(
+    pub async fn get_stream_by_position(
```
Comment (Li0k) [inline src/storage/src/hummock/sstable_store.rs:495]: The caller of `get_stream_by_position` must passed in a Some(block_idx), we can remove this Option
